rasa/core/policies/ted_policy.py

```diff
@@ -0,0 +1,636 @@
+        # set to 0 to turn off normalization
+        RANKING_LENGTH: 10,
+        # how similar the algorithm should try
+        # to make embedding vectors for correct labels
+        MAX_POS_SIM: 0.8,  # should be 0.0 < ... < 1.0 for 'cosine'
+        # maximum negative similarity for incorrect labels
+        MAX_NEG_SIM: -0.2,  # should be -1.0 < ... < 1.0 for 'cosine'
+        # the number of incorrect labels, the algorithm will minimize
+        # their similarity to the user input during training
+        USE_MAX_NEG_SIM: True,  # flag which loss function to use
+        # scale loss inverse proportionally to confidence of correct prediction
+        SCALE_LOSS: True,
+        # regularization
+        # the scale of regularization
+        REGULARIZATION_CONSTANT: 0.001,
+        # the scale of how important is to minimize the maximum similarity
+        # between embeddings of different labels
+        NEG_MARGIN_SCALE: 0.8,
+        # dropout rate for dial nn
+        DROPRATE_DIALOGUE: 0.1,
+        # dropout rate for bot nn
+        DROPRATE_LABEL: 0.0,
+        # dropout rate for attention
+        DROPRATE_ATTENTION: 0,
+        # visualization of accuracy
+        # how often calculate validation accuracy
+        EVAL_NUM_EPOCHS: 20,  # small values may hurt performance
+        # how many examples to use for hold out validation set
+        EVAL_NUM_EXAMPLES: 0,  # large values may hurt performance
+    }
+    # end default properties (DOC MARKER - don't remove)
+
+    @staticmethod
+    def _standard_featurizer(max_history: Optional[int] = None) -> TrackerFeaturizer:
+        if max_history is None:
+            return FullDialogueTrackerFeaturizer(LabelTokenizerSingleStateFeaturizer())
+        else:
+            return MaxHistoryTrackerFeaturizer(
+                LabelTokenizerSingleStateFeaturizer(), max_history=max_history
+            )
+
+    def __init__(
+        self,
+        featurizer: Optional[TrackerFeaturizer] = None,
+        priority: int = DEFAULT_POLICY_PRIORITY,
+        max_history: Optional[int] = None,
+        model: Optional[RasaModel] = None,
+        **kwargs: Dict[Text, Any],
+    ) -> None:
+        """Declare instant variables with default values"""
+
+        if not featurizer:
+            featurizer = self._standard_featurizer(max_history)
+
+        super().__init__(featurizer, priority)
+
+        self._load_params(**kwargs)
+
+        self.model = model
+
+        self._label_data = None
+        self.data_example = None
+
+    def _load_params(self, **kwargs: Dict[Text, Any]) -> None:
+        self.config = copy.deepcopy(self.defaults)
+        self.config.update(kwargs)
+
+        self.config = train_utils.check_deprecated_options(self.config)
+
+        self.config = train_utils.update_similarity_type(self.config)
+
+        if self.config[EVAL_NUM_EPOCHS] < 1:
+            self.config[EVAL_NUM_EPOCHS] = self.config[EPOCHS]
```

Comment:
I thought `-1` is a natural way to say "at the last epoch"

Reply:
ah I see. yes I think that makes sense. is that documented in the docs?

code wise I'd make that explicit as well, otherwise, this looks just like magic 😅 

```
            if self.config[EVAL_NUM_EPOCHS] == -1: 
              # magic value -1 is used to set evaluation to number of epochs
              self.config[EVAL_NUM_EPOCHS] = self.config[EPOCHS]
            elif self.config[EVAL_NUM_EPOCHS] < 1:
              raise ...
```